### PR TITLE
Add PostgresSQL module for Filebeat

### PR DESCRIPTION
**Pull request by kvch**
I have been working on a PostgreSQL module for Filebeat. Unfortunately, I got stuck at viewing logs in Kibana or maybe at indexing logs in Elasticsearch. 

The fields for PostgreSQL I have defined are visible in the index. When I test the the pipeline I have created using the Simulate API with the sample logs in the commit, the fields are parsed properly. However, when I read logs of PostgreSQL using Filebeat, it does not seem to work when viewed in Kibana.

This is what I see:
postgreslog

I would expect the logs to be properly parsed, but I don't think it is what I see. To be honest, I am not sure what to expect. Could you give me some pointers on how to go on?

Also I have collected a few TODOs, but please let me know if I missed something.

TODO
- [x] write proper documentation
- [x] gather more test logs
- [x] create visualizations in Kibana
- [ ] make checks green
- [x] add PR to changelog
- [x] add expected outputs

### CODE REVIEW

File: filebeat/module/postgresql/log/ingest/pipeline.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
+{
+  "description": "Pipeline for parsing PostgreSQL logs.",
+  "processors": [
+    {
+      "grok": {
+        "field": "message",
+        "ignore_missing": true,
+        "patterns": [
+          "%{LOCALDATETIME:postgresql.log.timestamp} %{WORD:postgresql.log.timeformat} \[%{NUMBER:postgresql.log.thread_id}\] %{USERNAME:postgresql.log.username}@%{HOSTNAME:postgresql.log.host} %{WORD:postgresql.log.level}:  ?%{GREEDYDATA:postgresql.log.message}",
+          "%{LOCALDATETIME:postgresql.log.timestamp} %{WORD:postgresql.log.timeformat} \[%{NUMBER:postgresql.log.thread_id}\] \[%{USERNAME:postgresql.username}\]@\[%{HOSTNAME:postgresql.log.host}\] %{WORD:postgresql.log.level}:  ?%{GREEDYDATA:postgresql.log.message}",
```
Comment (urso) [inline filebeat/module/postgresql/log/ingest/pipeline.json]: s/postgresql.username/postgreseql.log.username/ ?

File: filebeat/module/postgresql/log/ingest/pipeline.json [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+          "%{LOCALDATETIME:postgresql.log.timestamp} %{WORD:postgresql.log.timeformat} \[%{NUMBER:postgresql.log.thread_id}\] %{WORD:postgresql.log.level}:  ?%{GREEDYDATA:postgresql.log.message}"
```
Comment (urso) [inline filebeat/module/postgresql/log/ingest/pipeline.json]: I wonder if you need to use `\\` here. grok is basically build on regular expressions. Plus this file being json might might affect the `\` on read.

**Comment by kvch** [comment]
My initial problem was solved. Thank you for your help. :) I updated the TODO list based on the current state of the PR.

**Comment by tsg** [comment]
This is green enough, the remaining failure is Metricbeat on Windows.

**Comment by tsg** [comment]
@kvch It would be nice to also have a fileset for slow logs, like we have for MySQL. But totally fine to do that in a second PR.

**Comment by kvch** [comment]
@tsg I am trying to add slowlog parsing. It is not in a different fileset, because PostgreSQL puts slowlogs in the same file as the other logs by default.

Again the Grok parser has failed me. :D I have worked out a regex for multiline query statements using Kibana Dev Tools, but it does not seem to work with Filebeat.

**Comment by kvch** [comment]
The problem is fixed. Multiline configuration was missing from the module.

**Comment by tsg** [comment]
@kvch Something else I noticed, I think you didn't yet add `-expected.json` files. See for example, the one from the [Nginx module](https://github.com/elastic/beats/tree/master/filebeat/module/nginx/access/test). There's `test.log`, which contains a few sample log lines, and `test.log-expected.json` which contain the result after passing it through Elasticsearch. If you add such an `*-expected.json` file, the system tests will automatically check the results. See the responsible system test [here](https://github.com/elastic/beats/blob/master/filebeat/tests/system/test_modules.py#L118).

**Comment by kvch** [comment]
@tsg I will add it. Thanks.

**Comment by kvch** [comment]
I have added a new commit containing a test refactor.

**Comment by tsg** [comment]
Remaining test failure is not related. Merging.

File: filebeat/docs/modules/postgresql.asciidoc [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+This module comes with two sample dashboards.
+
+The first dashboard is for regulars logs.
+
+image::./images/filebeat-postgresql-overview.png[]
+
+The second one shows the slowlogs of PostgreSQL.
+
+image::./images/filebeat-postgresql-postgresql-overview.png[]
```
Comment (jpountz) [inline filebeat/docs/modules/postgresql.asciidoc]: it should be `images/filebeat-postgresql-slowlog-overview.png` I think

File: filebeat/module/postgresql/_meta/docs.asciidoc [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+This module comes with two sample dashboards.
+
+The first dashboard is for regulars logs.
+
+image::./images/filebeat-postgresql-overview.png[]
+
+The second one shows the slowlogs of PostgreSQL.
+
+image::./images/filebeat-postgresql-postgresql-overview.png[]
```
Comment (jpountz) [inline filebeat/module/postgresql/_meta/docs.asciidoc]: it should be `images/filebeat-postgresql-slowlog-overview.png` I think

File: filebeat/docs/modules/postgresql.asciidoc (hunk 3 above)
Comment (tsg) [inline filebeat/docs/modules/postgresql.asciidoc]: Should be fixed with: https://github.com/elastic/beats/pull/4843

**Comment by liuqian1990** [comment]
2017-12-27 11:19:07 GMT [30072]-[14] session=5a43112b.7578 user=postgres db=postgres app=[unknown] client=10.0.2.73 LOG:  duration: 1.229 ms  statement: 
	    SELECT COUNT(mode) AS count, mode
	      FROM pg_locks GROUP BY mode
	    UNION SELECT COUNT(*) AS count, 'waiting' AS mode
	      FROM pg_locks
	      WHERE granted is false;
	    
2017-12-27 11:19:07 GMT [30072]-[15] session=5a43112b.7578 user=postgres db=postgres app=[unknown] client=10.0.2.73 LOG:  duration: 0.334 ms  parse <unnamed>: SELECT pg_database_size($1) AS size;
2017-12-27 11:19:07 GMT [30072]-[16] session=5a43112b.7578 user=postgres db=postgres app=[unknown] client=10.0.2.73 LOG:  duration: 0.206 ms  bind <unnamed>: SELECT pg_database_size($1) AS size;
2017-12-27 11:19:07 GMT [30072]-[17] session=5a43112b.7578 user=postgres db=postgres app=[unknown] client=10.0.2.73 DETAIL:  parameters: $1 = 'postgres'
2017-12-27 11:19:07 GMT [30070]-[18] session=5a43112b.7576 user=postgres db=postgres app=[unknown] client=10.0.2.73 LOG:  duration: 52.563 ms  execute <unnamed>: SELECT pg_database_size($1) AS size;
2017-12-27 11:19:07 GMT [30070]-[19] session=5a43112b.7576 user=postgres db=postgres app=[unknown] client=10.0.2.73 DETAIL:  parameters: $1 = 'postgres'
2017-12-27 11:19:07 GMT [30070]-[20] session=5a43112b.7576 user=postgres db=postgres app=[unknown] client=10.0.2.73 LOG:  duration: 0.112 ms  parse <unnamed>: SELECT pg_database_size($1) AS size;
2017-12-27 11:19:07 GMT [30070]-[21] session=5a43112b.7576 user=postgres db=postgres app=[unknown] client=10.0.2.73 LOG:  duration: 0.165 ms  bind <unnamed>: SELECT pg_database_size($1) AS size;

###### How to resolve these logs, set up can not be resolved

**Comment by liuqian1990** [comment]
Provided Grok expressions do not match field value: [2017-12-27 13:54:07 GMT [30071]-[3741] session=5a43112b.7577 user=postgres db=postgres app=[unknown] client=10.0.2.73 LOG:  duration: 0.180 ms  bind <unnamed>: SELECT pg_database_size($1) AS size;]

**Comment by ruflin** [comment]
@liuqian1990 Could you please open a new topic on discuss for further discussions as this is a closed PR? https://discuss.elastic.co/c/beats/filebeat Please provide there your config and the filebeat version you are using + the logs as you did above.

**Pull request merged**